ospackage/recipes/artifact_repository.go

```diff
@@ -0,0 +1,221 @@
+	if testStorageClient != nil {
+		return testStorageClient, nil
+	}
+	return storage.NewClient(ctx)
+}
+
+func newHTTPClient() *http.Client {
+	if testHTTPClient != nil {
+		return testHTTPClient
+	}
+	return &http.Client{}
+}
+
+// FetchArtifacts takes in a slice of artifacs and dowloads them into the specified directory,
+// Returns a map of artifact names to their new locations on the local disk.
+func FetchArtifacts(ctx context.Context, artifacts []*osconfigpb.SoftwareRecipe_Artifact, directory string) (map[string]string, error) {
+	localNames := make(map[string]string)
+
+	for _, a := range artifacts {
+		path, err := fetchArtifact(ctx, a, directory)
+		if err != nil {
+			return nil, err
+		}
+		localNames[a.Id] = path
+	}
+
+	return localNames, nil
+}
+
+type objectLocation struct {
+	bucket string
+	path string
+	generation string
+}
+
+func tryGcsRegex(r *regexp.Regexp, url string) (objectLocation, bool) {
+	matches := gsHTTPRegex1.FindStringSubmatch(url)
+	if len(matches) == 3 {
+		return objectLocation{bucket: matches[1], path: matches[2]}, true
+	}
+	if len(matches) == 4 {
+		return objectLocation{bucket: matches[1], path: matches[2], generation: matches[3]}, true
+	}
+	return objectLocation{}, false
+}
+
+func tryParseGcsURL(url string) (objectLocation, bool) {
+	if loc, ok := tryGcsRegex(gsHTTPRegex1, url); ok {
+		return loc, true
+	}
+	if loc, ok = tryGcsRegex(gsHTTPRegex2, url); ok {
+		return loc, true
+	}
+	if loc, ok = tryGcsRegex(gsHTTPRegex2, url); ok {
+		return loc, true
+	}
+	return objectLocation{}, false
+}
+
+func fetchArtifact(ctx context.Context, artifact *osconfigpb.SoftwareRecipe_Artifact, directory string) (string, error) {
+	path := path.Join(directory, artifact.Id)
+	uri, err := url.Parse(artifact.Uri)
+	if err != nil {
+		return "", fmt.Errorf("Could not parse url %q for artifact %q", artifact.Uri, artifact.Id)
+	}
+	scheme := strings.ToLower(uri.Scheme)
+
+	switch scheme {
+	case "gcs":
+		loc, ok := tryGcsRegex(gsRegex, artifact.Uri)
+		if !ok {
+			return "", fmt.Errorf("Could not parse gs url %q for artifact %q", artifact.Uri, artifact.Id)
+		}
+		err := fetchFromGCS(ctx, artifact, loc, path)
+		if err != nil {
+			return "", err
+		}
+	case "https", "http":
+		loc, ok  := tryParseGcsURL(artifact.Uri)
+
+		if ok {
+			err = fetchFromGCS(ctx, artifact, loc, path)
+			if err != nil {
+				return "", err
+			}
+		} else {
+			err := fetchViaHTTP(ctx, artifact, path)
+			if err != nil {
+				return "", err
+			}
+		}
+	default:
+		return "", fmt.Errorf("protocol %q found in artifact %q not supported", scheme, artifact.Id)
+	}
+
+	return path, nil
+}
+
+func fetchFromGCS(ctx context.Context, artifact *osconfigpb.SoftwareRecipe_Artifact, loc objectLocation, path string) error {
+	client, err := newStorageClient(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to create storage client: %v", err)
+	}
+	defer client.Close()
+
+	oh := client.Bucket(uri.Hostname()).Object(uri.Path)
```

Comment:
you take in `loc objectLocation` but you're still parsing the unknown variable `uri` here. run `go vet && go build`

Reply:
I changed it to using the right variables.